crypto/dilithium/p_dilithium3.c

```diff
@@ -9,117 +9,300 @@
+    OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATON_NOT_INITIALIZED);
+    return 0;
+    }
 
-  DILITHIUM3_KEY *key = ctx->pkey->pkey.ptr;
+  NISTDSA_KEY *key = ctx->pkey->pkey.nistdsa_key;
 
-  if (siglen != DILITHIUM3_SIGNATURE_BYTES ||
-      ml_dsa_65_verify(tbs, tbslen, sig, siglen, NULL, 0, key->pub) != 0) {
+  if (siglen != nistdsa->signature_len ||
+      nistdsa->method->verify(tbs, tbslen, sig, siglen,
+                              NULL, 0, key->public_key) != 0) {
     OPENSSL_PUT_ERROR(EVP, EVP_R_INVALID_SIGNATURE);
+    return 0;
+      }
+
+  return 1;
+}
+
+// This function sets nistdsa parameters defined by |nid| in |pkey|.
+// If |pkey| already has a public key set, this public key is preserved.
+int EVP_PKEY_nistdsa_set_params(EVP_PKEY *pkey, int nid) {
+  const NISTDSA *nistdsa = SIG_find_dsa_by_nid(nid);
+
+  if (nistdsa == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_UNSUPPORTED_ALGORITHM);
+    return 0;
+  }
+
+  // if the public key has already been set either by EVP_parse_public_key or
+  // some other method that returns a PKEY without setting params, then
+  // we preserve that PKEY and just populate the params
+  if (pkey->pkey.nistdsa_key != NULL) {
+    pkey->pkey.nistdsa_key->nistdsa = nistdsa;
+    return 1;
+  }
+
+  evp_pkey_set_method(pkey, &nistdsa_asn1_meth);
+
+  NISTDSA_KEY *key = NISTDSA_KEY_new();
+  if (key == NULL) {
+    // NISTDSA_KEY_new sets the appropriate error.
     return 0;
   }
+
+  key->nistdsa = nistdsa;
+  pkey->pkey.nistdsa_key = key;
+
   return 1;
 }
 
-const EVP_PKEY_METHOD dilithium3_pkey_meth = {
-    EVP_PKEY_DILITHIUM3,
-    NULL /* init */,
-    NULL /* copy */,
-    NULL /* cleanup */,
-    pkey_dilithium3_keygen,
-    NULL /* sign_init */,
-    NULL /* sign */,
-    pkey_dilithium3_sign_message,
-    NULL /* verify_init */,
-    NULL /* verify */,
-    pkey_dilithium3_verify_message,
-    NULL /* verify_recover */,
-    NULL /* encrypt */,
-    NULL /* decrypt */,
-    NULL /* derive */,
-    NULL /* paramgen */,
-    NULL /* ctrl */,
-    NULL /* ctrl_str */,
-    NULL /* keygen deterministic */,
-    NULL /* encapsulate deterministic */,
-    NULL /* encapsulate */,
-    NULL /* decapsulate */,
+// Takes an EVP_PKEY_CTX object |ctx| and sets nistdsa parameters defined
+// by |nid|
+int EVP_PKEY_CTX_nistdsa_set_params(EVP_PKEY_CTX *ctx, int nid) {
+  if (ctx == NULL || ctx->data == NULL) {
+    OPENSSL_PUT_ERROR(EVP, ERR_R_PASSED_NULL_PARAMETER);
+    return 0;
+  }
+
+  // It's not allowed to change context parameters if
+  // a PKEY is already associated with the context.
+  if (ctx->pkey != NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_INVALID_OPERATION);
+    return 0;
+  }
+
+  const NISTDSA *nistdsa = SIG_find_dsa_by_nid(nid);
+  if (nistdsa == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_UNSUPPORTED_ALGORITHM);
+    return 0;
+  }
+
+  NISTDSA_PKEY_CTX *dctx = ctx->data;
+  dctx->nistdsa = nistdsa;
+
+  return 1;
+}
+
+const EVP_PKEY_METHOD nistdsa_pkey_meth = {
```

Comment:
You'll have to use `DEFINE_METHOD_FUNCTION` for the FIPS build.

Reply:
Yes, I'll make that change when we make the move to the `fipsmodule` with related changes around moving these methods from `non_fips_pkey_evp_methods`. I envision this within the same PR that removes the `enable_dilithium` flag to clean up all those ifdefs